### PR TITLE
Improve CI performance

### PR DESCRIPTION
**Pull request by breeswish**
## TL;DR

This PR eliminates PR's CI time from [1.5](https://circleci.com/gh/pingcap/tikv/11179) ~ [2.5 hours](https://circleci.com/gh/pingcap/tikv/11205) to less than [30 minutes](https://circleci.com/gh/pingcap/tikv/11240)!

### Prevent Cache Bloating (-25 min)

A fresh build of current TiKV only generates 3GB cache, whose downloading time is only 5 minutes. But latest cache size before I manually cleared bloated to [19GB](https://circleci.com/gh/pingcap/tikv/11205), whose downloading time was 30 minutes and uploading time was 50 minutes. This is huge.

So by preventing cache bloating, we can save a lot of CI time.

This PR permanently prevents cache bloating by limiting each build to either "load cache", or "save cache", but not both, more specifically:

- For non-master branches, like PRs, it always load (a clean) cache from master. It will not upload cache so that uploading time can be eliminated. (for a 3GB cache, uploading time cost another 5 minutes).

- For master branches, it never loads cache and always do fresh builds. After successful build, cache will be uploaded and will be used for other PRs. Notice that this ensures that cache is fresh and will not bloat, so it is fast to upload and also fast to download. This of course slows down the CI time (still notably faster then current master), but we actually don't care about how slow master builds.

### No Cache Uploading for non-master (-50 min)

As described previously, uploading cache also costs a lot of time. I have done the following two tests:

test 1: load latest PR's cache (5 min), build PR's change (5 min), upload PR's cache (5 min)
test 2: load master's cache (5 min), build PR's change (5 min)

Test 1 cost more time because latest PR's cache does NOT help eliminate build time compared to master's cache, while uploading will cost extra time.

So for non-master branches, cache uploading is completely removed.

### Disable Code Coverage for non-master (-30 min)

Currently we don't respect code coverage in PRs. Also according to a [tarpaulin's bug](https://github.com/xd009642/tarpaulin/issues/119), it cannot just replace unit tests. So code coverage is removed when building PRs. It is preserved for master branch so that we can still learn code coverage.

When the issue is fixed, we can switch back using tarpaulin in PRs and remove tests so that we can have both tests and code coverages in PRs while not doubling the test time.

## Statistics

After this PR is merged, PR's building time will be improved to about [30 minutes](https://circleci.com/gh/pingcap/tikv/11240), consists of:

- Install System Dependencies (2 min)
- Load master's ~/.cargo & ~/.local cache (0.5 min)
- Load master's ./target cache (5 min)
- Build (3 ~ 8 min)
- Run test (18 min)

Master branch's build time will be about 1.5 hour (+ coverage 40 min, + upload 5 min).

### CODE REVIEW

**Comment by breeswish** [comment]
@siddontang @Hoverbear @overvenus PTAL

**Comment by breeswish** [comment]
Now circle CI runs even faster than our Jenkins build 😂😂

**Comment by siddontang** [comment]
Cooooooooooooool 👍

**Comment by codeworm96** [comment]
Amazing

File: .circleci/config.yml [hunk 1]
```diff
@@ -32,17 +32,19 @@ jobs:
...
-            - v2-local-{{ .Branch }}-{{ .Environment.CIRCLE_SHA1 }}
-            - v2-local-{{ .Branch }}
-            - v2-local
-      - restore_cache:
-          name: Restoring ~/.cargo Cache
-          keys:
-            - v2-cargo-{{ .Branch }}-{{ checksum "Cargo.lock" }}
-            - v2-cargo-{{ .Branch }}
-            - v2-cargo
+            - v3-cargo-master
```
Comment (overvenus) [inline .circleci/config.yml]: Will it choose the most recent cache that starts with `v3-cargo-master`?

Comment (breeswish) [reply .circleci/config.yml]: Because PR's build will not upload valid cache (see readme idea 2). So for individual PRs, they receive snapshot from latest master, hoping for some usefulness.
Master does not need to download cache, but it is not possible to write conditional statement here, so leave it.

File: .circleci/config.yml [hunk 2]
```diff
@@ -76,41 +78,72 @@ jobs:
...
-            - ~/.local
-            - ~/.rustup
+            - ~/tikv_cache_cargo
       - run:
           name: Testing
           command: make trace_test
           no_output_timeout: 1800s
       - run:
-          name: Calculating code coverage
+          name: Calculating code coverage (master branch only)
```
Comment (Hoverbear) [inline .circleci/config.yml]: I guess this means no more reports per PR.

Comment (breeswish) [reply .circleci/config.yml]: yes. Once tarpaulin's bug is solved we can add code coverage back. Currently we are not very care about the code coverage in fact so temporarily removed.

File: .circleci/config.yml (hunk 1 above)
Comment (overvenus) [reply .circleci/config.yml]: Why not specify master's git hash? 

```
git rev-parse master
```

Comment (breeswish) [reply .circleci/config.yml]: First, you can't write anything dynamic here, so using a non const value is nearly impossible. (Notice that environment variable is also useless because in `restore_cache`, it only supports built-in environment variables and context variables).

Second, CircleCI matches cache by prefix and if there are multiple matches a latest one is used. So this is enough for us.

Comment (breeswish) [reply .circleci/config.yml]: I found out that I misunderstood your comment yesterday LOL the comment above should be the right answer. Yes, CircleCI will match to the most recent cache in that prefix, it is a documented behaviour.
